src/bp/nlu-core/intents/oos-intent-classfier.ts

```diff
@@ -0,0 +1,245 @@
+
+    const model = await svm.train([...in_ctx_scope_points, ...oos_points], trainingOptions, progress)
+    return model
+  }
+
+  private async _trainInScopeSvm(
+    trainInput: TrainInput,
+    noneIntent: Omit<Intent<Utterance>, 'contexts'>,
+    progress: (p: number) => void
+  ): Promise<string> {
+    const baseIntentClf = new BaseIntentClassifier(this.tools)
+    const noneUtts = noneIntent.utterances.filter(u => u.tokens.filter(t => t.isWord).length >= 3)
+    const trainableIntents = trainInput.intents.filter(
+      i => i.name !== NONE_INTENT && i.utterances.length >= MIN_NB_UTTERANCES
+    )
+    const nAvgUtts = Math.ceil(_.meanBy(trainableIntents, i => i.utterances.length))
+
+    const lo = this.tools.seededLodashProvider.getSeededLodash()
+
+    const intents: Intent<Utterance>[] = [
+      ...trainInput.intents,
+      {
+        name: NONE_INTENT,
+        utterances: lo
+          .chain(noneUtts)
+          .shuffle()
+          .take(nAvgUtts * 2.5) // undescriptible magic n, no sens to extract constant
+          .value(),
+        contexts: [...trainInput.intents[0].contexts],
+        slot_definitions: []
+      }
+    ]
+
+    await baseIntentClf.train({ ...trainInput, intents }, progress)
+    return baseIntentClf.serialize()
+  }
+
+  private getVocab(utts: Utterance[]) {
+    return _.flatMap(utts, u => u.tokens.map(t => t.toString({ lowerCase: true })))
+  }
+
+  public serialize(): string {
+    if (!this.model) {
+      throw new Error('Intent classifier must be trained before calling serialize')
+    }
+    return JSON.stringify(this.model)
+  }
+
+  public load(serialized: string): void {
+    const model: Model = JSON.parse(serialized) // TODO: validate input
+
+    const { oosSvmModel, baseIntentClfModel, trainingVocab } = model
+
+    const baseIntentClf = new BaseIntentClassifier(this.tools)
+    baseIntentClf.load(baseIntentClfModel)
+
+    this.predictors = {
+      oosSvm: oosSvmModel ? new this.tools.mlToolkit.SVM.Predictor(oosSvmModel) : undefined,
+      baseIntentClf,
+      trainingVocab
+    }
+  }
+
+  public async predict(utterance: Utterance): Promise<NoneableIntentPredictions> {
+    if (!this.predictors) {
+      throw new Error('Intent classifier must be either trained or a model must be loaded before calling predict')
+    }
+
+    const { oosSvm, baseIntentClf, trainingVocab } = this.predictors
+
+    const intentPredictions = await baseIntentClf.predict(utterance)
+
+    let oosPrediction = 0
+    if (oosSvm) {
+      const feats = getUtteranceFeatures(utterance, trainingVocab)
+      try {
+        const preds = await oosSvm.predict(feats)
+        oosPrediction =
+          _.chain(preds)
+            .filter(p => p.label.startsWith('out'))
+            .maxBy(p => p.confidence)
+            .value()?.confidence || 0
+      } catch (err) {}
+    }
+
+    // TODO: proceed to election between none intent and oos, remove none intent and make sure confidences sum to 1.
```

Comment:
Can you refresh my memory a bit, where is that election done now ? or is it that we don't have any

Reply:
We pick between none intent and oos field in the legacy election pipeline. I eventually want to move this election logic inside this box. 

As we've already discussed, the state of the flow can't help us choose between the two options so there's no need for delaying the decision.